server/controllers/profilephoto.js

```diff
@@ -0,0 +1,99 @@
+    const userId = req.user.id;
+
+    const profile = await Profile.findOne({ userId });
+    if (!profile) {
+        res.status(404);
+        throw new Error("Please create your profile first");
+    }
+    const s3 = new S3({
+        region: process.env.AWS_REGION,
+        accessKeyId: process.env.AWS__ACCESS_KEY,
+        secretAccessKey: process.env.AWS_SECRET_KEY,
+    })
+    if (profile.photoUrl) {
+        const pathname = profile.photoUrl.substring(48);
+        const deleteParams = {
+            Bucket: process.env.AWS_BUCKET,
+            Key: pathname,
+        }
+        await s3.deleteObject(deleteParams, (err) => {
+            if (err) {
+                res.status(400);
+                throw new Error('Error occured while trying to delete previous file in S3 bucket', err);
+            }
+        });
+    }
+    const uploadParams = {
+        Bucket: process.env.AWS_BUCKET,
+        Body: fs.createReadStream(req.file.path),
+        Key: profile.userId + (`${req.file.originalname}`),
+        ACL: 'public-read'
+    };
+    s3.upload(uploadParams, (err, data) => {
+        if (err) {
+            res.status(403).send('Error occured while trying to upload to S3 bucket');
+        }
+        if (data) {
+            fs.unlinkSync(req.file.path);
+            profile.photoUrl = data.Location;
+            profile
+                .save()
+                .then(profile => {
+                    res.json({ message: 'Upload Photo successfully', profile });
+                })
+                .catch(err => {
+                    res.status(400).send('Error occured while trying to save to DB');
+                });
+        }
+    });
+
+});
+
+// @route get /photourl
+// @desc Save user photo to local storage
+// @access Private
+exports.downloadPhoto = asyncHandler(async (req, res) => {
+    const userId = req.user.id;
+
+    const profile = await Profile.findOne({ userId });
+    if (!profile) {
+        res.status(404);
+        throw new Error("No profile to check");
+    }
+    const s3 = new S3({
+        region: process.env.AWS_REGION,
+        accessKeyId: process.env.AWS__ACCESS_KEY,
+        secretAccessKey: process.env.AWS_SECRET_KEY,
+    })
+    if (profile.photoUrl) {
+        const downloadParams = {
+            Bucket: process.env.AWS_BUCKET,
+            Key: profile.photoUrl.substring(48),
+        }
+        await s3.getObject(downloadParams,
+            function (error, data) {
+                if (error) {
+                    res.json("Failed to retrieve an object: ");
+                } else {
+                    fs.writeFile("../client/src/Images/" + profile.photoUrl.substring(72), data.Body, (err) => {
+                        if (err)
+                            res.json('Error occured while writing file');
+                        else {
+                            res.json("File written successfully");
+                        }
+                    });

```

Comment:
This only works because we host the server on the same machine as our client in development! We should send the file in a response instead (this [answer](https://stackoverflow.com/a/17818492) gives an example of how to do it in `Node.js`, you can use a similar method with `s3.getObject().createReadStream()` 👍🏼 )

Reply:
addressed